src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Taxation\Applicator;
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
+use Sylius\Component\Order\Factory\AdjustmentFactoryInterface;
+use Sylius\Component\Taxation\Calculator\CalculatorInterface;
+use Sylius\Component\Taxation\Resolver\TaxRateResolverInterface;
+
+/**
+ * @author Mark McKelvie <mark.mckelvie@reiss.com>
+ */
+class OrderItemUnitsTaxesApplicator implements OrderTaxesApplicatorInterface
+{
+    /**
+     * @var CalculatorInterface
+     */
+    private $calculator;
+
+    /**
+     * @var AdjustmentFactoryInterface
+     */
+    private $adjustmentFactory;
+
+    /**
+     * @var TaxRateResolverInterface
+     */
+    private $taxRateResolver;
+
+    /**
+     * @param CalculatorInterface $calculator
+     * @param AdjustmentFactoryInterface $adjustmentFactory
+     * @param TaxRateResolverInterface $taxRateResolver
+     */
+    public function __construct(CalculatorInterface $calculator, AdjustmentFactoryInterface $adjustmentFactory, TaxRateResolverInterface $taxRateResolver)
+    {
+        $this->calculator = $calculator;
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->taxRateResolver = $taxRateResolver;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order, ZoneInterface $zone)
+    {
+        foreach ($order->getItemUnits() as $unit) {
+            $item = $unit->getOrderItem();
+            $taxRate = $this->taxRateResolver->resolve($item->getVariant(), ['zone' => $zone]);
```

Comment:
hmmm....

It might be more efficient to have 2 loops, first loops around orderItems so you can resolve the tax rate just once per `OrderItem`, then an inner loop which calculates and applies it for each unit in the item?